src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javax.crypto.spec;
+
+import javax.crypto.SecretKey;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Parameters for the combined Extract-Only, Expand-Only, or Extract-then-Expand operations of the
+ * HMAC-based Key Derivation Function (HKDF). The HKDF function is defined in <a
+ * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+ *
+ * @since 23
+ */
+public interface HKDFParameterSpec extends KDFParameterSpec {
+
+    /**
+     * This builder helps with the mutation required by the {@code Extract} scenario.
+     */
+    final class Builder {
+
+        Extract extract = null;
```

Comment:
No need to store an `extract` field. Just create one and return it in `extractOnly`.

Reply:
This field is useless. Since the only place to create a builder is `return new Builder().createExtract()` in the `extract()` method, the field is always non null. But then when `exactOnly` is called you create a new one and return it. And when `andExpand` is call, you only check if it's null (which never is).